### ISSUE TITLE
SVGO v4

### ISSUE THREAD

**Issue by TrySound** [svg/svgo#1429]
Just a list of possible changes

- replace convertPathData applyTransforms option with applyTransforms plugin
- throw errors instead of putting into result.error
- replace JSAPI class with set of utilities

**Comment by strarsis** [comment]
Would it make sense to add the txml parser PR to the milestone, too?

**Comment by vanillajonathan** [comment]
Maybe [TypeScript](https://www.typescriptlang.org/)?

**Comment by TrySound** [comment]
Maybe. We already started adding jsdoc for ts.

**Comment by TrySound** [comment]
Started covering plugins with tsdoc. Quite handy.
https://github.com/svg/svgo/pull/1527

**Comment by septs** [comment]
Will this project eventually migrate to TypeScript only? or (most files)

**Comment by TrySound** [comment]
Probably no, I'd like to avoid build step. And tsdoc looks like good enough.

**Comment by vanillajonathan** [comment]
With tsdoc/esdoc/jsdoc you might want to generate HTML documentation which you can host on GitHub pages.

**Comment by oscarotero** [comment]
Typescript or, at least, ES modules would be great. I created a fork to provide support for Deno some time ago (https://github.com/lumeland/svgo) but it's hard to maintain and update every new version. If the SVGO core used ESM it would be much more easy to adapt for Deno.

**Comment by TrySound** [comment]
Thank you for this! Not sure about timing. Will try to start working on v3 before the end of the year.

**Comment by TrySound** [comment]
Btw @oscarotero svgo provides esm module for browsers
https://unpkg.com/browse/svgo@2.7.0/dist/svgo.browser.js

The entry point is this module https://github.com/svg/svgo/blob/master/lib/svgo.js

Instead of https://github.com/svg/svgo/blob/master/lib/svgo-node.js used for node.

Can this help with deno?

**Comment by oscarotero** [comment]
Yes, It's the same entry point that I've used in the Deno fork https://github.com/lumeland/svgo/blob/deno/mod.js
But the browser dist stores the SVG in `globalThis` (or window), and to use it in Deno it must be imported as module.

**Comment by TrySound** [comment]
SVGs are not stored in globalThis. Just some dependencies rely on globalThis to cache polyfills.

**Comment by oscarotero** [comment]
Sorry, you're right. I didn't know SVGO exports ES modules. Just tested it and seems to work fine. Thanks!

I'd love to have some official Deno support (to take advantage of types for configuration and stuff) but, for now, the browser dist just works.

**Comment by vanillajonathan** [comment]
If you decide not to use TypeScript you might be interested to generate TypeScript defintion files (`*.d.ts`).

https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html
https://github.com/microsoft/dts-gen

**Comment by TrySound** [comment]
That's not a problem. I'll publish types when whole project migrate to TSDoc.

**Comment by vanillajonathan** [comment]
> That's not a problem. I'll publish types when whole project migrate to TSDoc.

Maybe you could generate HTML docs and host it on GitHub Pages then also?

**Comment by TrySound** [comment]
@vanillajonathan You already wrote about this. Types are not ready yet.

**Comment by vanillajonathan** [comment]
I meant when you publish the types and when the project has finished its migration to TSDoc.

**Comment by amareshsm** [comment]
Hey  @TrySound,
 I would like to contribute to `replace default params object with default params syntax in destructuring ` change  Can you give some examples for this?
 
https://github.com/svg/svgo/blob/a9834efa1603198ae606765292215462ae193d0d/lib/svgo.js#L16-L19

  something like this?
 ```
 const optimize = (input:string='', config: { [key: string]: string; } | null = null) => { 
   if (config == null) { 
     config = {}; 
   } 
 
 ```

**Comment by JoKalliauer** [comment]
[svgo v3.0 is released](https://github.com/svg/svgo/releases). Is this issue still up to date?

**Comment by KrasnayaPloshchad** [comment]
Seen from [this repo](https://github.com/gildas-lormeau/mhtml-to-html), I think it would be nice if SVGO will be available on Deno, this will give users more choice to run SVGO.

**Comment by KTibow** [comment]
Closing this issue as SVGO v3 has been out for a while!
